lib/src/react_client/js_backed_map.dart

```diff
@@ -0,0 +1,139 @@
+
+  factory JsBackedMap.fromJs(JsMap other) {
+    // todo optimize by eliminating JsBackedMap instance creation
+    return new JsBackedMap.from(new JsBackedMap.backedBy(other));
+  }
+
+  static Map<K, V> copyToDart<K, V>(JsMap jsObject) {
+    return new Map<K, V>.from(new JsBackedMap.backedBy(jsObject));
+  }
+
+  // these checks moved to asserts for better inlining...
+  // todo see if we can keep toString() behavior of keys without asserts without breaking map behavior? probably not
+  bool _isValidKey(Object key) => key == null || key is String;
+
+  // ----------------------------------
+  // Core overrides
+  // ----------------------------------
+
+  @override
+  V operator [](Object key) {
+//    if (!_isValidKey(key)) return null;
+//    assert(_isValidKey(key));
+
+    return js_util.getProperty(jsObject, key);
+  }
+
+  @override
+  void operator []=(K key, V value) {
+    js_util.setProperty(jsObject, key, value);
+  }
+
+  // todo this cast seems to work in Dart 1 without much overhead in dart2js, but may break in DDC/Dartium and may also break reified types
+  @override
+  Iterable<K> get keys => _Object.keys(jsObject) as List<K>;
+
+  @override
+  V remove(Object key) {
+//    assert(_isValidKey(key));
+//    if (!_isValidKey(key)) return null;
+    final value = this[key];
+    _Reflect.deleteProperty(jsObject, key);
+    return value;
+  }
+
+  @override
+  void clear() {
+    for (var key in keys) {
+      _Reflect.deleteProperty(jsObject, key);
+    }
+  }
+
+  // ----------------------------------
+  // Optimized overrides
+  // ----------------------------------
+
+  @override
+  void addAll(Map<K, V> other) {
+    // Don't check against JsBackedMap<V> since it's less optimized in dart2js,
+    // and we can assume that any JsBackedMap that is also a Map<String, V>
+    // is also a JsBackedMap<V>. TODO validate in Dart 2 and create dart-lang/sdk bug for this
+    if (other is JsBackedMap) {
+      // This cast is necessary due to type inference not working
+      // properly without the generic parameter, and has no
+      // overhead in dart2js
+      _Object.assign(jsObject, (other as JsBackedMap).jsObject);
+    } else {
+      super.addAll(other);
+    }
+  }
+
+  @override
+  bool containsKey(Object key) {
+//    if (!_isValidKey(key)) return false;
+//    assert(_isValidKey(key));
+    return js_util.hasProperty(jsObject, key);
+  }
+
+  // todo this cast seems to work in Dart 1 without much overhead in dart2js, but may break in DDC/Dartium and may also break reified types
+  @override
+  Iterable<V> get values => _Object.values(jsObject) as List<V>;
+
+// TODO not sure if overriding makes sense; check to see which is more optimal for smaller maps
+//  @override
+//  bool containsValue(Object value) => values.contains(value);
+
+// overridden for to avoid cast in case the implementation of keys changes
+//  @override
+//  int get length => _Object.keys(jsObject).length;
```

Comment:
from line 99 to here all can probably go, unless the worry of implementation change is still relevant then id probably un-comment it out.